### DISCUSSION TITLE
A possible more flexible design of Labels for better online reconstruction support in different vendors

### DISCUSSION THREAD

**Discussion by zhengliuX** [pulseq/pulseq#76] [Ideas]
Hi all,
    
I want to discuss about a more flexible design of Labels in Pulseq that may have the potential to support online reconstruction in different vendors if necessary. For some brand new sequence designs, of course we need to do the reconstruction offline. But I guess in some cases, using vendors' online reconstruction is possible and needed. Pulseq provides some generic labels to support online reconstruction(including Gadgetron I suppose), but the design of labels that tagged the meta info of the ADC must have something that is verdor-specific. In those cases, online reconstruction is impossible.     

So I am thinking, to support different design of "labels" in different vendors, we could add a customizable configuration file (maybe just a txt file or jsom), like:
```json
[
    "LabelExt1",
    "LabelExt2",
    "LabelExt3"
]
```

the getSupportedLabels function would pre-define a set of generic labels, also it allows reading the configuration file to extend the list of labels based on the generic ones. Also the C++ interface has a similar logic. By designing such a configuration file (which should be provided by vendors), developers could add flexible labels to their seuqences based on the machines they are running. Of course, if the seq file is sent to a different vendor, which does not have the label configuration file, the label would just be seen as something like "Unknown", would not cause scanning failure  
<br/>
    This is just a primary thought of the labels. I don't think this is urgent. But I am willing to talk about this.

Best,
Zheng

**Comment by m-a-x-i-m-z** [comment]
Theoretically it is already like that, the labels are not fixed in the file format, they are easily extendable. The interpreter is supposed to ignore labels that are unknown to it without producing any errors. This capability we actually have to test, but in theory we are totally flexible already. 

What is poorly defined though at the moment is how the scan metadata is defined, on Siemens these are for example SeqExpo data (and few other data structures). Here I believe we could use quite some improvement in the Pulseq file format.

**Reply by zhengliuX** [reply]
Sorry, I am not familiar with the scan metadata in Siemens, is it the file storing info like patient parameters, or the protocol parameters, for example, TE and TR when scanning a product sequence.

**Reply by QingpingChen1201** [reply]
Hi,
for Siemens, the scan metadata is the data header of the acquired data, including patient parameters, protocol parameters (TE, TR, etc), and others. By using the LABELs in Pulseq, we could modify some of the data header information, which may provide accurate k-space information to aid in online reconstruction.
Best regards,
Qingping
